[Uyen Ngo] iP

uyencfi opened the pull request:
## Duke
> Because why not have
> a task manager using
> your own CLI?

What Duke can do:
- [x] save your tasks
- [x] mark them as done

What Duke ~~will not do~~ cannot do yet:
- order tasks by importance
- send reminders

Life of a Duke user:
1. Start Duke
2. Type away!
    * hit enter on a super long command but made a typo? Just be patient and type again :relieved:
3. Stop Duke.

__iP may have broken your heart :broken_heart: (it certainly broke mine)__
### *but Duke can heal it :gift_heart:*

Have a [sneak peek](https://www.youtube.com/watch?v=dQw4w9WgXcQ) at the code behind:
```java
public class Duke {
    public static void main(String[] args) {
        new Duke().run();
    }
    // ...
}
```

Commit:
Add Gradle support

uyencfi committed:
Add skeletal implementation to echo commands

uyencfi committed:
Store user inputs and display them

uyencfi committed:
Implement mark as done

uyencfi committed:
Add support for tracking three types of tasks

uyencfi committed:
Change 'Task' to abstract class

uyencfi committed:
Fix leading whitespace included in task name

uyencfi committed:
Add text UI testing

uyencfi committed:
Add total number of tasks to task list description

+ Update corresponding test cases

uyencfi committed:
Add error handling

uyencfi committed:
Add error handling for 'done' task out of range

uyencfi committed:
Improve error message for better clarity

uyencfi committed:
Fix bug: command 'bye bye' terminates programme

uyencfi committed:
Update tests to include tests for error

uyencfi committed:
Add support for deleting tasks

uyencfi committed:
Refactor: reduce code duplicate

uyencfi committed:
Add tests for deleting a task

uyencfi committed:
Add clean up method after programme ends

uyencfi committed:
Add persistent storage for tasks

uyencfi committed:
Add basic support for date data

uyencfi committed:
Add date and time support for deadline tasks

uyencfi committed:
Add start and end time support for event tasks

uyencfi committed:
Merge branch 'branch-Level-7'

uyencfi committed:
Merge branch 'branch-Level-8'

uyencfi committed:
Fix incomplete merge (?)

uyencfi committed:
Refactor for more OOP

To more closely conform to OOP, let's structure the code as
* Ui: user interaction (read input, print)
* Storage: load tasks and save tasks to/from the file
* Parser: parse text input and pass to Command
* Command & subclasses: execute commands, operate on task list and
produce a CommandResult
* CommandResult: contain operation result to report to user
* TaskList: contain the task list

In essence, file DukeBot.java is divided into the above classes.

uyencfi committed:
Update textUi tests

uyencfi committed:
Organize classes into packages

uyencfi committed:
Fix typo in output message

uyencfi committed:
Refactor: clean up package

uyencfi committed:
Add unit tests for duke.tasks package

uyencfi committed:
Add JavaDoc comments

uyencfi committed:
Tidy up: improve method name, add annotations

uyencfi committed:
Add search task functionality

uyencfi committed:
Add unit tests for search task functionality

uyencfi committed:
Merge branch 'branch-A-JavaDoc'

uyencfi committed:
Merge branch 'branch-A-CodingStandard'

uyencfi committed:
Merge branch 'branch-Level-9'

uyencfi committed:
Merge branch 'add-gradle-support'

uyencfi committed:
Add missing JavaDoc

Mistake spotted thanks to Gradle build JavaDoc.

uyencfi committed:
Add support for Checkstyle

uyencfi committed:
Refactor tests to comply with CheckStyle

uyencfi committed:
Refactor main code to comply to CheckStyle

uyencfi committed:
Handle unknown task type when loading data

Since the user can modify the file duke.txt which holds task data,
incorrect data may be entered. Let's handle all unknown task types by
throwing an error and creating a blank task list.

uyencfi committed:
[WIP] Implement a working GUI for Duke

uyencfi committed:
Improve dialog appearance, prevent truncating long text

uyencfi committed:
Update project structure to reflect GUI classes

Now that we have a GUI, we should re-organize the classes appropriately.

Inside "root" package `duke`, let's
- Remove text-based UI methods from Duke.java and put them in a new
class called TextDuke
- Rename the text UI class from `Ui` to `TextUi` for clear distinction
with the new GUI
- Put controller class `MainWindow` and `DialogBox` inside package `ui`
- Group packages `commands`, `parser`, `tasks` under a new package `logic`
to bundle together the inner workings of Duke
- Update unit test classes accordingly

uyencfi committed:
Tidy up miscellaneous nits

uyencfi committed:
Update app entry point for Gradle

uyencfi committed:
Merge branch 'branch-Level-10'

uyencfi committed:
Fix typo in Delete and Done command execution messages

uyencfi committed:
Fix Gradle build (?)

uyencfi committed:
Handle program exit

uyencfi committed:
Add CSS styling to JavaFX parts

uyencfi committed:
[WIP] Use basic regex to parse input

uyencfi committed:
Improve find task: multiple keywords, case insensitive

GabrielWLM commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.io.IOException;
+
+import duke.logic.Duke;
+import duke.ui.MainWindow;
+import javafx.application.Application;
+import javafx.fxml.FXMLLoader;
+import javafx.scene.Scene;
+import javafx.scene.layout.AnchorPane;
+import javafx.stage.Stage;
```
Perhaps you could split this group of import statements into smaller groups? I noticed similar sizable groups of import statements in several other classes.

uyencfi replied:
Thanks for the review! I configured my IntelliJ import statement organization according to the [AB3 convention](https://se-education.org/guides/tutorials/intellijCodeStyle.html#tweak-import-order), but I'll look into splitting them if it makes them more readable.

GabrielWLM commented on src/main/java/duke/TextDuke.java:
```diff
@@ -0,0 +1,77 @@
+        } catch (IOException e) {
+            textUi.showSavingError(e.getMessage());
+        }
+        textUi.showGoodbye();
+        textUi.close();
+        System.exit(0);
```
Any reason why you chose not to use a `finally` block?

uyencfi replied:
Oops never noticed this, I guess it's because the app doesn't exit after catching the error so the subsequent code will be executed regardless. But I'll add a `finally` block for clarity!

GabrielWLM commented on src/main/java/duke/logic/Duke.java:
```diff
@@ -0,0 +1,58 @@
+        return d;
+    }
+
+    /**
+     * Processes the user input and obtains the response.
+     *
+     * @param userCommandText The user input.
+     * @return A string describing the program's response to the user input.
+     */
+    public String getResponse(String userCommandText) {
+        Command command = parser.parse(userCommandText);
+        command.setTaskList(taskList);
+        CommandResult result = command.execute();
+        return result.getMessage();
+    }
```
Perhaps you could mention what this function returns in the summary of the method description?

I noticed a similar trend in some of the method descriptions. Is there a reason you prefer describing what the method is intended to be used for in another part of the code instead of what the method does and returns?

uyencfi replied:
I don't think the module's Java coding standard really specify what the header comment should contain (do correct me if I'm wrong!) so I followed how the Java API writes comment, which is just a short description of what the method does. And personally I think it would be verbose to include the return value, since there is a `@return` clause dedicated to that purpose already. But I suppose I can check with prof on this issue! Thanks for the input.

Dr-Octavius commented on src/main/java/duke/TextDuke.java:
```diff
@@ -31,15 +33,15 @@ private void run() {
             taskList = new TaskList();
         }
-        ui.showWelcome();
+        textUi.showWelcome();
```
Minor point, but maybe just leave this as "welcome" instead of "showWelcome"?

uyencfi replied:
I wanted to have `show` in the method name to indicate an action, will change to `showWelcomeMessage` to be more specific.

Dr-Octavius commented on src/main/java/Event.java:
```diff
@@ -1,11 +1,20 @@
         this.at = at;
     }
 
+    public String formatForSave() {
```
Maybe "formatSave" would be the same?

uyencfi replied:
Thanks for the suggestion! I couldn't think of a more concise name for this method so I had to put `formatForSave` 😅

Dr-Octavius reviewed (commented):
Overall, great use of OOP, apart from minor nomenclature (naming) issues. Otherwise, LGTM! Keep up the good work 😊

GabrielWLM commented on src/main/java/duke/logic/parser/Parser.java:
```diff
@@ -0,0 +1,202 @@
+     * @return The command based on the user input.
+     */
+    private Command parseTokens(String[] tokens) {
+        switch (tokens[0]) {
+        case "bye":
+            return handleBye(tokens);
+        case "list":
+            return handleList(tokens);
+        case "done":
+            return handleDone(tokens);
+        case "delete":
+            return handleDelete(tokens);
+        case "todo":
+            return handleToDo(tokens);
+        case "deadline":
+            return handleDeadline(tokens);
+        case "event":
+            return handleEvent(tokens);
+        case "find":
+            return handleFind(tokens);
```
Should a `//Fallthrough` comment be added in these cases as there is no break statement?

uyencfi replied:
I'm not sure if it's needed here, because the function immediately returns in each case so there is no possibility of falling through to a different case.

GabrielWLM reviewed (commented):
Nice job! I like your organization and your code is very readable.

Besides a few comments, I think your code has minimal coding violations. Good Job!

rgbpokka commented on src/main/java/duke/TextDuke.java:
```diff
@@ -0,0 +1,77 @@
+    private void runCommandLoopUntilExitCommand() {
+        Command command;
+        do {
+            String userCommandText = textUi.getUserCommand();
```
I like how the methods are clearly named and easy to understand!

uyencfi replied:
Thanks for the review! All the best for your iP too 💪

rgbpokka commented on src/main/java/duke/logic/parser/Parser.java:
```diff
@@ -0,0 +1,202 @@
+     * @return The command based on the user input.
+     */
+    private Command parseTokens(String[] tokens) {
+        switch (tokens[0]) {
```
Good use of switch statement here!

rgbpokka commented on src/main/java/duke/logic/parser/Parser.java:
```diff
@@ -0,0 +1,202 @@
+     * @param tokens The tokenized input array.
+     * @return The prepared command.
+     */
+    private Command handleEvent(String[] tokens) {
```
Thorough handling of the different cases, nice!

rgbpokka reviewed (commented):
Overall the code adheres to the code quality guidelines and I've personally learned from reading your code, great job!

uyencfi committed:
Rename some methods following PR review comments

uyencfi committed:
Add assertions for assumption checking

There is no assert statements in the code. Even though all existing
errors have been handled by try-catch blocks, it does no  harm to add
some assertions to double check the code.

Let's
- assert that all TaskList operations receive a list index within bounds
- assert that Parser does not pass empty args to user input handlers
- assert that the task list that commands operate on is not null
- finally, add enable assertion config in Gradle

uyencfi committed:
Fix bug: Stream reusing fails unit test

Creating just one stream and use it inside multiple filter operations
will throw an error because the stream has already been operated upon
once already.

Let's stream-ify the array `keywords` each time it is used, i.e. inside
the lambda for filter.

Moral of the story: Sometimes rash optimization does more harm than good.
 And always run tests to detect regression.

uyencfi committed:
Remove magic literals

uyencfi committed:
Command parsing: Convert string comparison to enum

uyencfi committed:
Merge pull request #1 from uyencfi/branch-A-Assertions

Add assertions to check assumption in code

uyencfi committed:
Merge branch 'master' into branch-A-CodeQuality

Update branch due to merged PR in remote master branch.

uyencfi committed:
Merge branch 'master' into branch-A-Streams

Update branch due to merged PR in remote master.

uyencfi committed:
Merge pull request #2 from uyencfi/branch-A-CodeQuality

Extract magic strings and create an enum class for command parsing

uyencfi committed:
Merge branch 'master' into branch-A-Streams

uyencfi committed:
Merge pull request #3 from uyencfi/branch-A-Streams

Fix bug: Stream reusing fails unit test

uyencfi committed:
Set up CI with GitHub Actions

uyencfi committed:
Fix minor coding standard violations (to pass CI)

uyencfi committed:
Fix datetime formatter bug: Specify the locale

uyencfi committed:
[wip] Add support for updating tasks

uyencfi committed:
Add some unit tests for updating tasks

uyencfi committed:
Add some unit tests for Update command

uyencfi committed:
Move TaskStub to a nested class in TaskListTest

Putting TaskStub on the top level clutters the package and makes
it harder to identify test classes.

Let's move it to a nested class, putting it inside TaskListTest.
This makes sense because TaskListTest directly requires a stub for task,
and doing this clears up the package.

uyencfi committed:
Refactor ArgumentTokenizer into a utility class

uyencfi committed:
Add Javadoc, clean up code

uyencfi committed:
Merge pull request #4 from uyencfi/branch-C-Update

Add feature: Update tasks

uyencfi committed:
Switch to JavaFX 11 for JDK 11 compatibility

uyencfi committed:
Finalize features

uyencfi committed:
Add Ui screenshot

uyencfi committed:
Set theme jekyll-theme-minimal

uyencfi committed:
Merge remote-tracking branch 'origin/master'

uyencfi committed:
Add a brief User guide

uyencfi committed:
Fix test case

uyencfi committed:
Add meta info for GitHub Page

uyencfi committed:
Update _config.yml

uyencfi committed:
Add logo to display on GitHub Page

uyencfi committed:
Merge remote-tracking branch 'origin/master'

uyencfi committed:
Fix logo path

uyencfi committed:
Fix emoji cannot display on GitHub Page

Change from GitHub markdown syntax into HTML code

uyencfi committed:
Delete unnecessary line

uyencfi committed:
Move PageLogo to publishing source directory

uyencfi committed:
Update release version to v0.2
